types/xdg_shell/wlr_xdg_toplevel.c

```diff
@@ -65,49 +82,49 @@ void send_xdg_toplevel_configure(struct wlr_xdg_surface *surface,
 		struct wlr_xdg_surface_configure *configure) {
 	assert(surface->role == WLR_XDG_SURFACE_ROLE_TOPLEVEL);
 
-	configure->toplevel_state = malloc(sizeof(*configure->toplevel_state));
-	if (configure->toplevel_state == NULL) {
+	configure->toplevel_configure = malloc(sizeof(*configure->toplevel_configure));
+	if (configure->toplevel_configure == NULL) {
 		wlr_log(WLR_ERROR, "Allocation failed");
 		wl_resource_post_no_memory(surface->toplevel->resource);
 		return;
 	}
-	*configure->toplevel_state = surface->toplevel->server_pending;
+	*configure->toplevel_configure = surface->toplevel->scheduled;
 
 	struct wl_array states;
 	wl_array_init(&states);
-	if (surface->toplevel->server_pending.maximized) {
+	if (surface->toplevel->scheduled.maximized) {
 		uint32_t *s = wl_array_add(&states, sizeof(uint32_t));
```

Comment:
Unrelated note: seems like this code could be cleaner by filling a `uint32_t states[32]`, then creating a `wl_array` pointing to it. Something like:

```c
uint32_t states[32];
size_t states_len = 0;

if (scheduled.maximized) {
    states[states_len++] = XDG_TOPLEVEL_STATE_MAXIMIZED;
}
…

struct wl_array states_arr = {
    .data = states,
    .size = states_len * sizeof(uint32_t),
};
```

Reply:
Seems like a pretty unconventional way to create a wl_array, but sure, why not. Also avoids potential OOM errors.

Reply:
Yeah… Ideally `wl_array` could be marked `const` so that wayland-util functions like `wl_array_add` can't be called, but [it's a bit too late](https://gitlab.freedesktop.org/wayland/wayland/-/merge_requests/107). Here `wl_array` is just a read-only structure used by libwayland to send events.